### PR TITLE
PR ontram-pkg

### PR DESCRIPTION
**Pull request by AGoetschi**

### CODE REVIEW

File: R/fit_ontram.R [hunk 1]
```diff
@@ -78,13 +136,47 @@ fit_ontram <- function(model, history = FALSE, x_train = NULL,
...
       model_history$test_loss <- append(model_history$test_loss, test_loss)
+      if (early_stopping) {
+          if (model_history$test_loss[epo] <= current_min) {
+            current_min <- model_history$test_loss[epo]
+            n_worse <- 0
+            if (epo == model$epoch) {
+              model_history$epoch_best <- epo
+            }
+            if (save_best) {
+            save_model_ontram(model, filename = paste0(filepath, "best_model"))
```
Comment (LucasKookUZH) [inline R/fit_ontram.R:147]: Add file ending `.h5` to `"best_model"`.

File: R/fit_ontram.R [hunk 2]
```diff
@@ -6,18 +6,66 @@
...
+    stopifnot(min_delta >= 0)
+    if (is.null(y_test)) stop("`y_test` not found.")
+    if (!is.null(x_train)) {
+      if(is.null(x_test)) stop("`x_test` not found.")
+    }
+    if (!is.null(img_train)) {
+      if (is.null(img_test)) stop("`img_test` not found.")
+    }
+    if (save_best) {
+      if (is.null(filepath) || !dir.exists(filepath)) stop("`filepath` missing or does not exist.")
```
Comment (LucasKookUZH) [inline R/fit_ontram.R:64]: Maybe instead of stopping, `dir.create()` the missing directory?

File: R/methods.R [hunk 3]
```diff
@@ -88,35 +88,206 @@ plot.ontram_history <- function(object, col_train = "blue", col_test = "red", ..
...
+#' x_valid <- x_train[1:20,]
+#' y_valid <- y_train[1:20,]
+#' mo <- ontram_polr(x_dim = ncol(x_train), y_dim = ncol(y_train),
+#'                   method = "logit", n_batches = 10, epochs = 50)
+#' fit_ontram(mo, x_train = x_train, y_train = y_train)
+#' simulate(mo, nsim = 1, x = x_valid, y = y_valid)
+#' @export
+simulate.ontram <- function(object, nsim = 1, seed = NULL, x = NULL, y, im = NULL, ...) {
+  if (!exists(".Random.seed", envir = .GlobalEnv, inherits = FALSE))
+    runif(1)
```
Comment (LucasKook) [inline R/methods.R]: This `runif(1)` call looks a bit weird. I'd probably assign it to a variable? Is it just to get the `".Random.seed"` initialized?

File: R/methods.R [hunk 4]
```diff
@@ -88,35 +88,206 @@ plot.ontram_history <- function(object, col_train = "blue", col_test = "red", ..
...
+    set.seed(seed)
+    RNGstate <- structure(seed, kind = as.list(RNGkind()))
+    on.exit(assign(".Random.seed", R.seed, envir = .GlobalEnv))
+  }
+  pr <- predict(object, x = x, y = y, im = im)
+  ret <- apply(pr$pdf, 1, function(p) sample(length(p), nsim, prob = p, replace = TRUE))
+  if (nsim > 1) {
+    tmp <- vector(mode = "list", length = nsim)
+    for (i in 1:nsim) {
+      tmp[[i]] <- ordered(ret[i, ])
```
Comment (LucasKook) [inline R/methods.R]: Add the correct levels to the ordered outcome using `levels = ...`

File: R/save-load_ontram.R [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  } else {
+    mb <- NULL
+  }
+  if (file.exists(nm_eta)) {
+    me <- load_model_hdf5(nm_eta)
+  } else {
+    me <- NULL
+  }
+  ret <- append(rest, list(mod_baseline = mt, mod_shift = mb, mod_image = me,
+                           optimizer = tf$keras$optimizers$Adam(learning_rate = 0.001),
```
Comment (LucasKook) [inline R/save-load_ontram.R]: The learning rate should probably be written and read as well. The `distr` argument can remain `tf$sigmoid` for the time being.

File: README.md [hunk 6]
```diff
@@ -4,5 +4,5 @@ R package for fitting ordinal neural network transformation models
 Install the package from within `R` via
 
 ```
-remotes::install_github("LucasKookUZH/ontram-pkg")
+remotes::install_github("AGoetschi/ontram-pkg")
```
Comment (LucasKook) [inline README.md]: Please revert change.

File: R/save-load_history.R [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                                dimnames = list(c("train_loss", "test_loss"), NULL))),
+              file = filepath, sep = ",", row.names = TRUE, col.names = FALSE)
+  if (length(object) > 2) {
+    write.table(object$epoch_best, file = filepath, sep = ",",
+                row.names = "epoch_best", col.names = FALSE,
+                append = TRUE)
+  }
+}
+
+#' @rdname save_history_ontram
```
Comment (LucasKook) [inline R/save-load_history.R]: Remove/change `@rdname` field.

File: R/save-load_ontram.R [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+  save_model_hdf5(object$mod_baseline, nm_theta)
+  if (!is.null(object$mod_shift)) {
+    save_model_hdf5(object$mod_shift, nm_beta)
+  }
+  if (!is.null(object$mod_image)) {
+    save_model_hdf5(object$mod_image, nm_eta)
+  }
+}
+
+#' @rdname save_model_ontram
```
Comment (LucasKook) [inline R/save-load_ontram.R]: Change/remove `@rdname` field.

File: R/fit_ontram.R (hunk 2 above)
Comment (LucasKook) [inline R/fit_ontram.R]: Maybe instead of stopping, `dir.create()` the missing directory?

**Review by LucasKook — CHANGES_REQUESTED** [review]
Thanks for the PR! There are some minor comments in the review, please address.

File: R/save-load_ontram.R [hunk 9]
```diff
@@ -43,8 +42,11 @@ load_model_ontram <- function(filename, ...) {
     me <- NULL
   }
   ret <- append(rest, list(mod_baseline = mt, mod_shift = mb, mod_image = me,
-                           optimizer = tf$keras$optimizers$Adam(learning_rate = 0.001),
-                           distr = tf$sigmoid))
+                           optimizer = tf$keras$optimizers$Adam(learning_rate = rest$lr),
+                           distr = rest$distr))
   class(ret) <- "ontram"
+  if (ret$response_varying) {
+    class(ret) <- append("ontram_rv", class(ret))
```
Comment (LucasKook) [inline R/save-load_ontram.R:49]: `append()` adds to the end of a vector, however `"ontram_rv"` should be the first class.
```r
class(ret) <- c("ontram_rv", class(ret))
```
will fix this.

**Review by LucasKook — APPROVED** [review]
Thanks!
